api_design/DDS-1182-nonchucked_upload_api_design.md

```diff
@@ -0,0 +1,157 @@
+# DDS-1182 Non-chunked uploads
+
+## Logical View
+
+#### Background
+
+Chunked uploads are useful when dealing with large files. The overhead of the chunked
+upload process is not ideal for small files.
+
+#### Proposal
+
+Add a `non_chunked` boolean attribute to the `POST /projects/{id}/uploads` endpoint.
+When `non_chunked` is true, the returned payload will include a signed url for
+uploading data.
+
+For uploads created with a truthy `non_chunked` attribute, the payload for the
+`GET /uploads/{id}` endpoint will include a signed url for uploading data.
+
+When the client is finished uploading data, the `PUT /uploads/{id}/complete`
+endpoint must be called. The hash algorithm may need to be validated to work with
+the upload's storage provider. Once the upload has been completed, the payload for
+`GET /uploads/{id}` will no longer include a signed url for uploading.
+
+These changes should speed up client interactions by removing the need to call the
+`PUT /uploads/{id}/chunks` endpoint to create a signed url for uploading. Also,
+since the DDS API does not need to interact with an external service to create
+signed urls, latency introduced by the eventual consistency model is also removed.
+
+## Implementation View
+
+#### Summary of impacted APIs
+
+|Endpoint |Description |
+|---|---|
+| `POST /projects/{id}/uploads` | Initiate upload |
+| `GET /uploads/{id}` | View upload |
+| `PUT /uploads/{id}/complete` | Complete file upload |
+| `PUT /uploads/{id}/chunks` | Get pre-signed chunk URL |
+
+#### API Specification
+
+##### Initiate upload
+
+`POST /projects/{id}/uploads`
+
+###### Request Properties
+
+- *name (string, required)* - The name of the client file to upload.
+- *content_type (string, optional)* - Valid content type per [media types](https://en.wikipedia.org/wiki/Internet_media_type).
+- *size (number, required)* - The size in bytes of entire file (computed by client).
+- *storage_provider.id (string, optional)* - The unique id for a storage provider.
+- *non_chunked (boolean, optional)* - When true, a signed upload url is returned with the payload
```

Comment:
since it is optional, it should state that the default is false

Reply:
also, POST /projects/{id}/uploads and GET /uploads/{id} should include an example of a non_chunked: false response to make it clear what to expect.

Reply:
> since it is optional, it should state that the default is false

Done.

Reply:
> also, POST /projects/{id}/uploads and GET /uploads/{id} should include an example of a non_chunked: false response to make it clear what to expect.

I'd rather keep the examples in this design doc to what is changing/being added. I'll add in a sentence or two explaining that the proposed changes are backwards compatible and responses will remain consistent with the previous version when `non_chunked` is not set.

Reply:
@dmlond  Ok, I added a few words about backwards compatibility. Check out my changes and let me know what you think. Thanks!